vertx-web-client/src/main/java/io/vertx/ext/web/client/impl/HttpRequestImpl.java

```diff
@@ -226,9 +232,21 @@ public void sendForm(MultiMap body, Handler<AsyncResult<HttpResponse<T>>> handle
   public void sendMultipartForm(MultipartForm body, Handler<AsyncResult<HttpResponse<T>>> handler) {
     send("multipart/form-data", body, handler);
   }
+  
+  protected void addOnContextCreated(Consumer<HttpContext> consumer) {
```

Comment:
@vietj I did a small change to HttpRequestImpl, allowing me to add attributes to HttpContext upfront, so I was then able to use the itnerceptorm mechanism.

I think this should be ok, do you agree?

Reply:
Hi @vietj,
I had a look at that branch (forgot to mention it, sorry) the interceptors, as implemented in there looks nicer to me but I still cannot see a way to map the request to the client that originated it.

It would work if we extend WebClientImpl instead of decorating it; but the, we will lose the one to many relations between the actual WebClient and the SessionAwareWebClient(s).

To allow which decorator we need to use for request/response processing we need some sort of attribute set on the request itself, before we return it to the caller, so that when he/she will call `request.send` we will e able to read that attribute back from our interceptor (both new or old style).

My implementation, allows you to set one or many `Consumer<HttpContext>` that will be invoked after the HttpContext is created an before interceptors will be invoked, so that you can add as many attributes as you want to the HttpContext and use them from the interceptors. 

Of course, I could be missing something... :)

Reply:
that's fine, after it's merged I'll have a look and see if I can make a refactor.

Reply:
(which is why testing matters so much :-) )